### ISSUE TITLE
Authentication broken again

### ISSUE THREAD

**Issue by pulsebreaker** [philippelt/netatmo-api-python#82]
Hi,

Something in the authentication is broken (again) it seems. I'm getting this since 12:05 this afternoon:

code=400, reason=, body=b'{"error":"invalid_grant"}'
Traceback (most recent call last):
  File "C:/Program Files/JetBrains/PyCharm 2023.3.2/plugins/python/helpers/pydev/pydevd.py", line 1534, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm 2023.3.2\plugins\python\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "F:\Programmeren\PyCharm\netatmo\netatmo_weatherstation-latest.py", line 17, in <module>
    ws = lnetatmo.WeatherStationData(authorization)
  File "C:\Users\signu\AppData\Local\Programs\Python\Python37-32\lib\site-packages\lnetatmo.py", line 282, in __init__
    self.getAuthToken = authData.accessToken
  File "C:\Users\signu\AppData\Local\Programs\Python\Python37-32\lib\site-packages\lnetatmo.py", line 181, in accessToken
    if self.expiration < time.time() : self.renew_token()
  File "C:\Users\signu\AppData\Local\Programs\Python\Python37-32\lib\site-packages\lnetatmo.py", line 192, in renew_token
    if self.refreshToken != resp['refresh_token']:
TypeError: 'NoneType' object is not subscriptable

I checked my app in Netatmo Connect, it's still there, nothing changed.

Kind regards,

Arno

**Comment by rvk01** [comment]
Are you working with the .netatmo.credentials file for authentication?
Is the content of that .netatmo.credentials valid?
And what time/date does that file have?

(For me it still works.)

**Comment by pulsebreaker** [comment]
I do use the file and the contents are still the same. I checked my script on two locations (NAS and local, both with the same credentials file) and both produce the same result. Weird. 

It still works for you so I will check my end again, thanks. 

Arno

________________________________
Van: rvk01 ***@***.***> 
verzonden: maandag 3 juni 2024 18:08 
Aan: philippelt/netatmo-api-python 
Cc: pulsebreaker; Author
Onderwerp: Re: [philippelt/netatmo-api-python] Authentication broken again (Issue #82)

Are you working with the .netatmo.credentials file for authentication?
Is the content of that .netatmo.credentials valid?
And what time/date does that file have?
(For me it still works.)

**Comment by rvk01** [comment]
I just got word from another user (with a raspberry pi reading out netatmo) where the Grafana view also isn't updated from around 12:00 today (last entries 11:45 CEST I see). So there might indeed be a problem. (I need to investigate this after diner ;) ).

**Comment by philippelt** [comment]
could you check that you are using the last version of the library (from repo, not yet from pip ) ?

**Comment by pulsebreaker** [comment]
Okay, my 'solution' didn't work, it just works 1 time, if I try for a
second time I get the same error.

I was using version 3.0.0. Updated it to 4.0.0, same error.

Then I downloaded the repo and did an overwrite of the lnetatmo.py file
with the one from the zipfile, same result:

code=400, reason=, body=b'{"error":"invalid_grant"}'
Traceback (most recent call last):
  File "C:/Program Files/JetBrains/PyCharm
2023.3.2/plugins/python/helpers/pydev/pydevd.py", line 1534, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm
2023.3.2\plugins\python\helpers\pydev\_pydev_imps\_pydev_execfile.py", line
18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "F:\Programmeren\PyCharm\netatmo\netatmo_weatherstation-latest.py",
line 17, in <module>
    ws = lnetatmo.WeatherStationData(authorization)
  File
"C:\Users\signu\AppData\Local\Programs\Python\Python37-32\lib\site-packages\lnetatmo.py",
line 413, in __init__
    self.getAuthToken = authData.accessToken
  File
"C:\Users\signu\AppData\Local\Programs\Python\Python37-32\lib\site-packages\lnetatmo.py",
line 240, in accessToken
    if self.expiration < time.time() : self.renew_token()
  File
"C:\Users\signu\AppData\Local\Programs\Python\Python37-32\lib\site-packages\lnetatmo.py",
line 251, in renew_token
    if self.refreshToken != resp['refresh_token']:
TypeError: 'NoneType' object is not subscriptable

If I manually generate a token on the Netatmo site and copy it in the
credentials files it works again but just 1 time.

Regards,

Arno


Op ma 3 jun 2024 om 19:18 schreef Philippe Larduinat <
***@***.***>:

**Comment by philippelt** [comment]
Sorry but I didn't updated the pypi version as I wanted to finish some refactoring.

Could you pick the library from github directly (4.1.0) ?

**Comment by rvk01** [comment]
Haha. The really weird part.... the installation that I mentioned (where it stopped working around 12:00 CEST) was running my own implementation in PHP. I don't use a webcreated token there but use 2FA for that (loging in and allowing access). It fails with invalid_grant directly trying to get a new refresh token.

My own installation here, with this netatmo-api-python code, still works fine, and still retrieves new access_tokens. I can see that from the time for the .netatmo file.

In my own implementation (at the remote site) I get a code via 2FA which I need to pass with grant_type' => "authorization_code". There it failes with {"error":"invalid_grant"}. But the code here doesn't use that (I think). So there are multiple issues at the end of netatmo.

**Comment by philippelt** [comment]
Many thanks @rvk01 for your feedback.

So it looks like Netatmo is playing again with authentication ! My various tasks are still working here but I don't know if it will last long !

Thanks for any news/feedback in this thread...

**Comment by pulsebreaker** [comment]
I directly downloaded the code from here, still the same issue.

It worked flawlessly for the last year or something until this afternoon so I suspected something changed or got broken at the Netatmo side.

The windmeter broke a couple of months ago also btw. I received no more winddata, only the battery status, time etc. and that was also in the official app. When it drained it's new batteries within a month I gave up on that.

Kinda done with the expensive unreliable stuff from Netatmo.

Regards,

Arno

**Comment by rvk01** [comment]
> It worked flawlessly for the last year or something until this afternoon so I suspected something changed or got broken at the Netatmo side.
I suspect that newly created access_tokens have problems with being passed with (any) grant_type.
Older tokens seem to work fine.
Maybe there was a problem around 12:00 and newly created tokens have that problem.

Authentication site is also kinda slow.
And now I sometimes get a "500 COMMON.SERVER ERROR" on https://api.netatmo.com/oauth2/authorize

Guess we need to wait a bit to see if it resolves itself.

**Comment by rvk01** [comment]
Aaaaah. I've got my own implementation working again :)
At the end it seems I used a client_id for doing OAuth2 which was different from the one for retrieving the tokens.

I'm not sure if it's the same problem in the code in this library (because it doesn't do OAuth2).

I did change another thing but I'm not sure if it could also be an issue.
When entering the app page on dev.netatmo.com, I saw that there is a (new) mandatory field "data protection officer name" & email. Those had asterisks (so mandatory) but where not filled in (they were not there when the app was registered). But because I filled them in before I saw the actual problem with the client_id, I'm not sure if it might be an issue.

Hopefully it keeps working here...

**Comment by pulsebreaker** [comment]
Just tried again, no changes. It did work, it doesn't work now.

Arno

Op ma 3 jun 2024 om 22:26 schreef rvk01 ***@***.***>:

**Comment by rvk01** [comment]
> Just tried again, no changes. It did work, it doesn't work now.

Did you add the data protection officer name & email in the app registration page?

I just tried it with freshly created token via the web interface and it worked here.
(used user root instead of normal pi to test)

**Comment by pulsebreaker** [comment]
I had both already filled in unfortunately.

Situation remains the same, if I refresh the token manually on the Netatmo
website and copy the new Refresh Token in the credentials file it will work
but just 1 time.

Op ma 3 jun 2024 om 23:08 schreef rvk01 ***@***.***>:

**Comment by rvk01** [comment]
> I had both already filled in unfortunately. Situation remains the same, if I refresh the token manually on the Netatmo website and copy the new Refresh Token in the credentials file it will work but just 1 time.

Did you check if the "CLIENT_ID" and "CLIENT_SECRET" are also exactly as stated on the App-page? For me, with the OAuth2 method, I used the wrong CLIENT_ID. In the past a different CLIENT_ID worked fine but it crapped out at 12:00 this afternoon.

**Comment by pulsebreaker** [comment]
Yes I did to be sure.

I copied the lnetatmo.py file from the github and pasted it directly in the
site-packages directory on my pc. This seemed to change something (I'm not
sure) because it looks like it works again. But when I update the
credentials file on my NAS, which is where I want it to run, it works just
1 time. After that the script also doesn't work on my pc anymore.

This tells me that the credentials file on the NAS isn't updated with the
new refresh token since this afternoon. It did before, strange.

Also funny, some dashboard data sometimes doesn't show up, but that's for
later.

It's late, I'm going to bed and continue tomorrow.

Thanks for all the help.

Arno

Op ma 3 jun 2024 om 23:41 schreef rvk01 ***@***.***>:

**Comment by rvk01** [comment]
> Yes I did to be sure. I copied the lnetatmo.py file from the github and pasted it directly in the site-packages directory on my pc. This seemed to change something (I'm not sure) because it looks like it works again. But when I update the credentials file on my NAS, which is where I want it to run, it works just 1 time.

That seems to indicate it is a local persmission problem.
I also copied the lnetatmo.py because it wasn't installed under my root user (just to test) and it worked. But I just copied it next to my .py file. So if it works there on the pc, then it should also work in the NAS.

Are you sure the correct library for lnetatmo.py is loaded on the NAS?
What are the permissions of the credential file?
Are they correct for the user which python script is running under?

I noticed the library only stores the REFRESH_TOKEN, not the ACCESS_TOKEN. The ACCESS_TOKEN is valid for about 3 hours. So the library always retrieves a new ACCESS_TOKEN when started. If the corresponding REFRESH_TOKEN (which is also renewed) can't be stored in the credential file, the next run will fail with an old REFRESH_TOKEN.

**Comment by pulsebreaker** [comment]
Problem solved!

You were right, it was an issue with not writing the new Refresh Token to the credentials file on the Synology NAS. The file permissions were still good (unchanged) so that was strange. 

I managed to get the newer lnetatmo.py file from here into the site-packages directory on the NAS (took some fiddling), checked all file permissions again and bingo, it started to work again.

I guess the version of lnetatmo I used was too old, something changed at the Netatmo site and it stopped working.

Thank you again for all the help.

**Comment by philippelt** [comment]
Good news @pulsebreaker , I am glad that authentication did not change again :wink:

**Comment by philippelt** [comment]
**BREAKING NEWS** ,

one user sent me today a response from Netatmo support about recent authentication issues. Here is a copy of this response :  

>
>Hello,
> 
>We just did a modification on the token retrieval process :
> 
>When you refreshed an access_token using the associated endpoint https://api.netatmo.com/oauth2/token, Netatmo servers responded with a couple of tokens : an access_token and a refresh_token. **If the previous access_token was still valid, the refresh_token value was never renewed**
> 
>Now, this behavior changed to be compliant with the recommendations of the RFC of the OAuth2 Authorization Framework (section 10.4) and improving the security of the data of our users
When refreshing tokens, access_token and refresh_token values will be automatically renewed and former tokens invalidated
>
>So, if you do not update and use the new refresh_token value when refreshing your access_token, your users will be disconnected after 3 hours and you will retrieve an “invalid_grant” error
To fix it, you need to update the tokens value as soon as you get the newly generated ones

Si finally there was an effective change in the Netatmo authentication service. I don't know if this was announced/documented anywhere but Netatmo has a long tradition of breaking things that works :smile: 

This can explain why we did have different behaviors exposed here. If you were using new token at each launch (like I do with cron starting my small monitor agent), you should not encounter problems. If you have monitor programs running permanently, that means calling Authentication only on startup, it will rely on management of access token expiration to get renewed.

Thus I have to check that, in that case, that the refresh token renewal is correctly handled.

**Comment by devMaFi** [comment]
On Monday at 12am my implementation stopped working. So I downloaded the latest version of the lnetamo.py file and give it a try...some issue as reported, one called worked, the following failed with the invalid grant error.

After changing to a credentials file with client id, client secret and refresh token I found at, that the refresh token is updated after each request.

Since I adapted this, the service is running continuously without issues.

May this helps the one or other

**Comment by LArtisanDuDev** [comment]
> **BREAKING NEWS** ,
> 
> one user sent me today a response from Netatmo support about recent authentication issues. Here is a copy of this response :
> 
> > Hello,
> > We just did a modification on the token retrieval process :
> > When you refreshed an access_token using the associated endpoint https://api.netatmo.com/oauth2/token, Netatmo servers responded with a couple of tokens : an access_token and a refresh_token. **If the previous access_token was still valid, the refresh_token value was never renewed**
> > Now, this behavior changed to be compliant with the recommendations of the RFC of the OAuth2 Authorization Framework (section 10.4) and improving the security of the data of our users
> > When refreshing tokens, access_token and refresh_token values will be automatically renewed and former tokens invalidated
> > So, if you do not update and use the new refresh_token value when refreshing your access_token, your users will be disconnected after 3 hours and you will retrieve an “invalid_grant” error
> > To fix it, you need to update the tokens value as soon as you get the newly generated ones
> 
> Si finally there was an effective change in the Netatmo authentication service. I don't know if this was announced/documented anywhere but Netatmo has a long tradition of breaking things that works 😄
> 
> This can explain why we did have different behaviors exposed here. If you were using new token at each launch (like I do with cron starting my small monitor agent), you should not encounter problems. If you have monitor programs running permanently, that means calling Authentication only on startup, it will rely on management of access token expiration to get renewed.
> 
> Thus I have to check that, in that case, that the refresh token renewal is correctly handled.

Thanks for these news.
I've done api call in C++ on ESP32 and my authent is broken since couple of days. Netatmo did it again !!

**Comment by rvk01** [comment]
The break from monday also had another factor. 
I wrongly used a different client_id with my access and refresh tokens. It was generated with the correct one but after that, the client_id wasn't checked any more (and I used another one).

As of monday 12:00 CEST netatmo implemented a change that client_id is also checked (and needs to be identical to the one with which the refresh_token was created). I found this out with my own PHP implementation.

@philippelt I still think the access_token and maybe time valid should be stored in de credentials file. access_token is normaly 3 hours valid and you don't need to throw it away each time you end the script. If you keep the script running for a long time, this doesn't matter but if you run the script every 10 minutes, there would be no need to get a new access_token. Netatmo doesn't penalize this yet but that could change in the future (getting new access_tokens while the previous one is still valid).

Only if using the access_token and getting an invalid error due to timeout of 3 hours, you can retrieve a new access_token and refresh_token. That's how I do it in PHP. I save the complete response from the retrieval of the access_token and reuse it until it errors out (after which I get a new one).

Maybe something for your todo-list 😁

**Comment by snikula** [comment]
Hi, im running simple test using cron, because my python-scrip which is used to retrieve my house temperature does not work after Netatmo change something? Test-script:

import lnetatmo from datetime import datetime
print(datetime.now()) f = open("/home/sami/test.txt", "w") f.write(str(datetime.now())) f.close()
authorization = lnetatmo.ClientAuth() devList = lnetatmo.WeatherStationData(authorization)
print (devList.lastData()['Ulkolämpötila']['Temperature'])

Done yesterday:
1. 07:27 authorize app after generating new Refresh Token (https://dev.netatmo.com/)
2. Writing new Refresh token in .netatmo.credentials (its writable to test.py:  -rwxrwxrwx 1 root root 189  6. 6. 07:28 .netatmo.credentials)
3. First run: 2024-06-06 07:29:38.117679
4. New Refresh token is writen to .netatmo.credentials in every test.py run
5. After ~3h it stops working
7. Trying to run manually to get error code, result:
code=400, reason=, body=b'{"error":"invalid_grant"}'
Traceback (most recent call last):
  File "/home/sami/test.py", line 15, in <module>
    devList = lnetatmo.WeatherStationData(authorization)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/sami/.local/lib/python3.11/site-packages/lnetatmo.py", line 413, in __init__
    self.getAuthToken = authData.accessToken
                        ^^^^^^^^^^^^^^^^^^^^
  File "/home/sami/.local/lib/python3.11/site-packages/lnetatmo.py", line 240, in accessToken
    if self.expiration < time.time() : self.renew_token()
                                       ^^^^^^^^^^^^^^^^^^
  File "/home/sami/.local/lib/python3.11/site-packages/lnetatmo.py", line 251, in renew_token
    if self.refreshToken != resp['refresh_token']:
                            ~~~~^^^^^^^^^^^^^^^^^
If I generate new Refresh token in dev.netatmo and writing it to .netatmo.credentials script will work again ~3h.

Any ideas?

**Comment by mika1337** [comment]
Hi all.

I have done some test in a fork of this library and a solution that seems to work is to:
- store the access_token in credentials file
- if the token hasn't expire, use directly the access_token
- upon token expiration, update access_token and refresh_token in the credentials file

**Comment by rvk01** [comment]
Yes, that's how I do it.

> * if the token hasn't expire, use directly the access_token

Only thing is that I don't store the expiration date. I just always try the access_token and if I get a error.code == 3 then the access_token is expired and I need to get a new one.

But I also have script 4.0.0 running here, and that one still works.

**Comment by rvk01** [comment]
> 1. 07:27 authorize app after generating new Refresh Token (https://dev.netatmo.com/)
> 2. Writing new Refresh token in .netatmo.credentials (its writable to test.py:  -rwxrwxrwx 1 root root 189  6. 6. 07:28 .netatmo.credentials)
> 3. First run: 2024-06-06 07:29:38.117679
> 4. New Refresh token is writen to .netatmo.credentials in every test.py run
> 5. After ~3h it stops working

I first thought it could be something with self.expiration (because there is potentially a problem on exact second of expiration). But that only happens if you keep your script running.

You start the script each time again and again. Then the access- and refresh-tokens should be refreshed each time.

- Please check if you are using the latest lnetatmo.py (the linenumbers in your error seem to be different from latest source).
- And also please check if you have the same CLIENT_ID in your credential file as in the web-site for the app at dev.netatmo.com.

**Comment by snikula** [comment]
Thanks for help!
CLIENT_ID is same
Where to find latest release? Im trying wget https://github.com/philippelt/netatmo-api-python/archive/v4.1.0.tar.gz, (from setup.py) ERROR 404: Not Found.  

> * Please check if you are using the latest lnetatmo.py (the linenumbers in your error seem to be different from latest source).
> * And also please check if you have the same CLIENT_ID in your credential file as in the web-site for the app at dev.netatmo.com.

**Comment by rvk01** [comment]
> Thanks for help! CLIENT_ID is same Where to find latest release? Im trying wget https://github.com/philippelt/netatmo-api-python/archive/v4.1.0.tar.gz, (from setup.py) ERROR 404: Not Found.

You can click code above at the left and click code again at the right (green button) then download as .zip.
It's this link
https://github.com/philippelt/netatmo-api-python/archive/refs/heads/master.zip

It will give you the latest from this repository.
(Not sure what the other link points at)

I'm not sure if pip also has the the latest.

**Comment by philippelt** [comment]
Hello all,

In fact, the netatmo change should not impact existing code. It's right that renewing access token at each run is more than strictly required but it should not create any problem.

I wonder if there can be an issue with the home relative path of the credential file.

for people who still have problems, would you please pick directly from github the last release 4.1.0 and add to your `ClientAuth` call the credentialFile parameter like the following example :  
```
authorization = lnetatmo.ClientAuth(credentialFile="/my/absolute/path/to/.netatmo.credentials")
```

**Comment by snikula** [comment]
> https://github.com/philippelt/netatmo-api-python/archive/refs/heads/master.zip
Thanks! Ill try this version.

**Comment by philippelt** [comment]
I understand your point @rvk01 but as Netatmo declared in their response, they want to **comply with the RFC of the OAuth2 Authorization Framework** and there is no such behavior described in the specs.

You can always loose an access token, if your program encounter a bug and fail for example. If you can't obtain a new token from the refresh token, you would be obliged to wait for 3 hours before being able to recover or restart from scratch by creating a new refresh token.

By the way, your suggestion to save the access token to be able to reuse it while it is still valid can be a good idea. I can't remember for the moment why I decided not to save it. I Think it was because I had to different servers asking for data with the same refresh token in the past (when refresh token did last forever) and I was afraid of side effects of access token invalidation by one server where the other was trying to use it.

I am not yet sure that our current problems are related to the access token. Strangely, I have been able to reproduce the problem using cron tasks and the solution was to explicitly mention the credential file path. It is like if `expanduser("~/.netatmo.credentials")` was no longer working for cron jobs. The current issue may be the consequence of python/OS upgrades (expanduser implementation changed in python 3.10.2) from our servers rather than Netatmo servers for which, as I said, the change should not break existing code.

In this case, I should try to use `Path.home()` and Path module methods instead of simply `os.path.expanduser()` the relative file path.

**Comment by rvk01** [comment]
> By the way, your suggestion to save the access token to be able to reuse it while it is still valid can be a good idea. I can't remember for the moment why I decided not to save it. I Think it was because I had to different servers asking for data with the same refresh token in the past (when refresh token did last forever) and I was afraid of side effects of access token invalidation by one server where the other was trying to use it.

No, server etc. were always the same. As I can remember you didn't implement this because you always had use-cases where the script was long-running. In which case the access_token was reused because it was in memory. You never had need for starting the script via cron every 10 minutes or so. That's why, I think, you didn't implement the saving of the access_token (we had a discussion about this in another issue in the past).

> I am not yet sure that our current problems are related to the access token. Strangely, I have been able to reproduce the problem using cron tasks and the solution was to explicitly mention the credential file path. It is like if `expanduser("~/.netatmo.credentials")` was no longer working for cron jobs.

For me it still works fine (python 3.11.2) as user pi.

BTW. I already mentioned, but there might be a small time critical issue when you have a long running script. You save self.expiration in the script as time+expires_in. But this could be a second or so off in which case the access_token was expired. For example if the accesstoken() is called one second before the time expires (but the server at netatmo sees it as such).

So you might want to build in a few seconds of margin in this line (substract 10 seconds from self.expiration for example):
` if self.expiration < time.time() : self.renew_token()`

Because if the access_token is expired... I also don't see any fallback after each communication to getting a new one in the code when the access_token is invalid... or am I missing something?

**Comment by pulsebreaker** [comment]
Following this discussion closely :-)

In my case the script on the NAS is started and ended every 15 minutes (because you get new data from Netatmo only every 10 minutes or so?), it updates the Refresh Token in the credentials file each run and is still working perfectly since 4 days ago.

**Comment by trygveasp** [comment]
I saw this issue first now when I wanted to report my findings. I was able to fix my code, but I have to make sure I read the REFRESH_TOKEN from the credential file and construct the ClientAuth object with refreshToken as an argument. 

```
        if os.path.exists(credentials):
            with open(credentials, mode="r") as fh:
                refreshToken = json.load(fh)["REFRESH_TOKEN"]
            authorization = lnetatmo.ClientAuth(refreshToken=refreshToken)
        else:
            authorization = lnetatmo.ClientAuth()
```

The problem is that unless this is specified it will always use the one from when lnetatmo was imported and that is normally not what you want if you like me have a real-time updating weather station display. So the lnetatmo code is not really broken, but I am not so sure the default refreshToken int ClientAuth should be the one from when the library was imported (_REFRESH_TOKEN)

**Comment by rvk01** [comment]
> I saw this issue first now when I wanted to report my findings. I was able to fix my code, but I have to make sure I read the REFRESH_TOKEN from the credential file and construct the ClientAuth object with refreshToken as an argument.

That's strange because that's exactly what the current code in the library does. Are you sure you have the latest version?

**Comment by trygveasp** [comment]
> > I saw this issue first now when I wanted to report my findings. I was able to fix my code, but I have to make sure I read the REFRESH_TOKEN from the credential file and construct the ClientAuth object with refreshToken as an argument.
> 
> That's strange because that's exactly what the current code in the library does. Are you sure you have the latest version? 
> 

I use 4.0.0 installed from pypi with pip. The problem with that version is that although the refresh token is renewed and the data is updated the first time, the default constructor always uses the original token from the import and thus throwing invalid grant.

**Comment by rvk01** [comment]
> I use 4.0.0 installed from pypi with pip. The problem with that version is that although the refresh token is renewed and the data is updated the first time, the default constructor always uses the original token from the import and thus throwing invalid grant.

That's strange. I also use 4.0.0 here (which is not the latest) and for me it always picks the correct one from the credential file. 

Are you sure there are no environment settings set for refresh_token. When they are, they are preferred over the credential file, I think.

You say 'from the import'. What do you mean by that. The refresh_token should (and is) always taken from the credential file (unless specifically passes in the function call, which shouldn't be necessary if the correct one is in the credential file.

So just set it manually the first time in the credential file (make sure the permissions are still correct) and call the function without parameters.

**Comment by trygveasp** [comment]
I use no enviroment variables any longer. And you are right, the first time the refresh token is read from the credential file in a variable which is set when the lnetatmo module is imported in my script. This variable is used as the default argument in the constructor and is never updated (at least in my case) even though the refresh token is updated  in the credential file.

**Comment by rvk01** [comment]
> This variable is used as the default argument in the constructor and is never updated (at least in my case) even though the refresh token is updated in the credential file.

It (self.refreshtoken) is updated in def renew_token(self) (when retrieving the access_tokeb). At least in the latest code. Also there on line 252 of version 4.0.0.

**Comment by trygveasp** [comment]
An example to demonstrate how it fails:

```
import lnetatmo
import time

def get_measurement():
    try:
        authorization = lnetatmo.ClientAuth()
        weather_data = lnetatmo.WeatherStationData(authorization)
    except:
        raise Exception

while True:
   print("get_measurement")
   get_measurement()
   time.sleep(1)
```
The first construction of ClientAuth is fine. The first call to WeatherStationData is fine but triggers an update of REFRESH_TOKEN and the ClientAuth object is updated accordingly and also the ~/.netatmo.credentials file. So far all good. However, in next iteration the ClientAuth still gets the REFRESH_TOKEN from the initial call to ClientAuth and the raise an invalid_grant exception. I guess many use this kind of workflow.

**Comment by rvk01** [comment]
> However, in next iteration the ClientAuth still gets the REFRESH_TOKEN from the initial call to ClientAuth and the raise an invalid_grant exception. I guess many use this kind of workflow.

How are you determining this? 
With a long running script (like you are showing) the internal self.refrestoken should be used. It shouldn't take any old REFRESH_TOKEN. Where in the code do you see that it takes the old one?

Anyway, for long running script you definitely should take the last version 4.1.0. Not 4.0.0. That version might not work for long running scripts. Only for short running cron scripts.

**Comment by trygveasp** [comment]
> > However, in next iteration the ClientAuth still gets the REFRESH_TOKEN from the initial call to ClientAuth and the raise an invalid_grant exception. I guess many use this kind of workflow.
> 
> How are you determining this? With a long running script (like you are showing) the internal self.refrestoken should be used (at least in latest 4.1.0). It shouldn't take any old REFRESH_TOKEN. Where in the code do you see that it takes the old one?
> 
> Anyway, for long running script you definitely should take the last version 4.1.0. Not 4.0.0. That version might not work for long running scripts. Only for short running cron scripts.

I know it because I print the params to postRequest and I see that refreshToken never changes even though it has been renewed, updated and the crediential file is updated:
```
def postRequest(topic, url, params=None, timeout=10):
    if PYTHON3:
        req = urllib.request.Request(url)
        if params:
            req.add_header("Content-Type","application/x-www-form-urlencoded;charset=utf-8")
            if "access_token" in params:
                req.add_header("Authorization","Bearer %s" % params.pop("access_token"))
            params = urllib.parse.urlencode(params).encode('utf-8')
        try:
            print("params", params)
            resp = urllib.request.urlopen(req, params, timeout=timeout) if params else urllib.request.urlopen(req, timeout=timeout)
```
My code was running fine with 4.0.0 in my long running script until last week when netatmo changed their API. I can test 4.1.0, but there is no package on pypi and no tag or release in this repo.... So where do I find 4.1.0?

**Comment by rvk01** [comment]
> My code was running fine with 4.0.0 in my long running script until last week when netatmo changed their API. I can test 4.1.0, but there is no package on pypi and no tag or release in this repo.... So where do I find 4.1.0?

You can click code above at the left and click code again at the right (green button) then download as .zip.
It's this link
https://github.com/philippelt/netatmo-api-python/archive/refs/heads/master.zip

It will give you the latest from this repository.
I'm not sure why pip doesn't have the latest.

**Comment by snikula** [comment]
> > Thanks for help! CLIENT_ID is same Where to find latest release? Im trying wget https://github.com/philippelt/netatmo-api-python/archive/v4.1.0.tar.gz, (from setup.py) ERROR 404: Not Found.
>  
> I'm not sure if pip also has the the latest.

Thanks! 
My script (and test-script) is now running with 4.1.0 with no errors.

**Comment by philippelt** [comment]
OK, there are troubles on Netatmo servers. To clarify the situation, I created a small script that just authent periodically (every 30 min) with always the same credential file.

Today at 5:02 UTC, I log this error :
```
Your current token scope do not allow access to Weather station
```
Yesterday at 7:02 UTC
```
code=400, reason=, body=b'{"error":"invalid_grant"}'
```
At jun 6, 5:02 UTC
```
Your current token scope do not allow access to Weather station
```

And the script is currently still running fine every 30 min without any change whether in the code or the .netatmo.credentials file.

Obviously there is an intermittent problem on Netatmo side. I think that the 4.1.0 version of the library should work without problem.

I will keep my script running to continue monitoring of Netatmo servers.

**Comment by trygveasp** [comment]
> You can click code above at the left and click code again at the right (green button) then download as .zip. It's this link https://github.com/philippelt/netatmo-api-python/archive/refs/heads/master.zip
> 
> It will give you the latest from this repository. I'm not sure why pip doesn't have the latest.

I can confirm that both my original code and the test I provided works fine with master! Would be nice to have a release of 4.1.0 on github/pip instead of downloading master;-)

**Comment by buldre** [comment]
I got the same issues as described above yesterday and today with 4.0.0 on a Raspberry Pi 4 running Python 3.11. I refreshed the Refresh Token and changed the permission on the credentials file. The file was then updated on each code start with a new Refresh Token. Then it worked once with a new Refresh Token, and now it always fails, ant the credentials file does not update.

I then tried 4.1.0 which also fails and does not update the file.

Here are two runs with 4.0.0 and 4.1.0:
Owner  : user
Group  : user
View   : Anybody
Change : Only owner
Execute: Nobody
Note = New Refresh Token. File NOT overwritten.
>>> %Run fnct_verify_Netatmo_connected.py
Printing Summary of Netatmo connection verification - Ver. 0
2024-06-21 15:12:19

code=400, reason=, body=b'{"error":"invalid_grant"}'
ERROR:root:Exception occurred
Traceback (most recent call last):
  File "/home/user/PythonScripts/ECfan/fnct_verify_Netatmo_connected.py", line 29, in verifyConnected
    weatherData = lnetatmo.WeatherStationData(authorization)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/PythonScripts/.venv/lib/python3.11/site-packages/lnetatmo.py", line 413, in __init__
    self.getAuthToken = authData.accessToken
                        ^^^^^^^^^^^^^^^^^^^^
  File "/home/user/PythonScripts/.venv/lib/python3.11/site-packages/lnetatmo.py", line 240, in accessToken
    if self.expiration < time.time() : self.renew_token()
                                       ^^^^^^^^^^^^^^^^^^
  File "/home/user/PythonScripts/.venv/lib/python3.11/site-packages/lnetatmo.py", line 251, in renew_token
    if self.refreshToken != resp['refresh_token']:
                            ~~~~^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable
e: 'NoneType' object is not subscriptable
>>> 

Permissions = Same as above
Note = Same Refresh Token as earlier today which has not been overwritten. File NOT overwritten in this test either
>>> %Run fnct_verify_Netatmo_connected.py
Printing Summary of Netatmo connection verification - Ver. 0
2024-06-21 20:01:43

auth...= <lnetatmo.ClientAuth object at 0x7fb0638c10>
code=400, reason=, body=b'{"error":"invalid_grant"}'
ERROR:root:Exception occurred
Traceback (most recent call last):
  File "/home/user/PythonScripts/ECfan/fnct_verify_Netatmo_connected.py", line 29, in verifyConnected
    weatherData = lnetatmo.WeatherStationData(authorization)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/PythonScripts/.venv/lib/python3.11/site-packages/lnetatmo.py", line 416, in __init__
    self.getAuthToken = authData.accessToken
                        ^^^^^^^^^^^^^^^^^^^^
  File "/home/user/PythonScripts/.venv/lib/python3.11/site-packages/lnetatmo.py", line 241, in accessToken
    if self.expiration < time.time() : self.renew_token()
                                       ^^^^^^^^^^^^^^^^^^
  File "/home/user/PythonScripts/.venv/lib/python3.11/site-packages/lnetatmo.py", line 252, in renew_token
    if self.refreshToken != resp['refresh_token']:
                            ~~~~^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable
e: 'NoneType' object is not subscriptable
>>>

**Comment by rvk01** [comment]
> I refreshed the Refresh Token and changed the permission on the credentials file. The file was then updated on each code start with a new Refresh Token. Then it worked once with a new Refresh Token, and now it always fails, ant the credentials file does not update.

The library could handle the grant error somewhat better (instead of crashing).

Did you create the refresh-token via the web interface of dev.netatmo.com? Did you also check if the client_id is exactly correct? If it's not it'll also crash with grant error.

**Comment by buldre** [comment]
Yes on dev.netatmo.com.
I updated my first comment, so you'll see I made a careful token transfer;-)

**Comment by rvk01** [comment]
> My client_ID and client_secret has not changed at all since I first made OAuth work with lnetatmo.

No, but netatmo made a change a few weeks ago to test for correct client-id. While you could use a wrong one in the past, you can't anymore and the client-id need to match the one on the dev page exactly. The secret however shouldn't be needed. Do you specify the secret and id somewhere in the code? If you do, that might override something.

**Comment by buldre** [comment]
No, they are only in the credentials file and are exact copies from the web page

**Comment by rvk01** [comment]
> I noted however, when I clicked to copy the tokens on my Netatmo page in Safari and pasted them into a Notes document on my Mac that some capital letters pasted as small caps, so beware! Better to read the Tokens off the Netatmo page and take care to check capital O versus zeros and small caps L versus capital I.

It's strange that some have this problem and others not (although I'm not sure if there are still others with problems)

BTW. The refresh-token and client-id do not contain O or L characters. It's base64 so only contains 0 to 9 and a to f (with a pipe | in the middle for the refresh). All lowercase (although that might not matter). The client-secret does contain other characters. (I now see the secret does get used)

**Comment by buldre** [comment]
In fact, my client_ID and secret have not changed since I created the App, and that was long before the OAuth. Maybe I should?

**Comment by rvk01** [comment]
You can always try to create a new app. Also note that "data protection officer" name and email is now mandatory (this also changed recently). Maybe that could be a reason.

**Comment by buldre** [comment]
What is "data protection officer", and is it used in the library, and do I need to use it in my code?

**Comment by rvk01** [comment]
No, it's not something you need in your code. But on dev.netatmo.com in your account, it has appeared as an input field (next to the app name etc) which has an asterisks next to it, so it's a mandatory field. You need to fill it in before you can change anything there.

**Comment by buldre** [comment]
Thanks! I'll try tomorrow. Bed time in home of yr now :-)

**Comment by buldre** [comment]
On first start the new app runs. I will wait at least three hours before I stop and try to restart.

Here is what I did:
I created a new app which is configured as a copy of my first with a different name.
The Client_ID is new and, still 24 characters.
The Client_Secret is new and, 44 characters. The old had 37.
The first part of the Access Token is the same as the first part of the Refresh Token, 24 characters, and equal to the old.
The second part of the Access Token is new and holds 32 characters like the old.
The second part of the Refresh Token is new and holds 32 characters like the old.

The credentials file uses TAB in front of each credential. 
Owner : user
Group : user
View : Anybody
Change : Anybody
Execute: Nobody

The second part of the Refresh Token was overwritten on start of the code - which calls the authentication only once.
This command runs every 10 minutes: weatherData = lnetatmo.WeatherStationData(authorization)
The code imports lnetatmo version 4.1.0.
Python 3.11 on Raspberry Pi 4

Crossing fingers!!!

**Comment by buldre** [comment]
It seems to work with short on/off intervals now, as expected.
:-)

**Comment by ultima-originem** [comment]
My code stopped working on 2024-06-04 with the dreaded "invalid_grant" error. I haven't been successful in solving this problem. Here's what I did:

- updated lnetatmo from 4.0.0 to 4.1.0
- registered a new app, with all new id, secret and token. 
- updated lnetatmo.credentials with the new info
- ran my code from a 5m crontab

and here's what happened:

- the authorization immediately started working again and the code ran for the first time at 22:40
- the code stopped running at 02:00 with the same "invalid_grant"error

Can anyone tell me what to do to resolve this issue? I've read the complete tread but am running out of ideas.

**Comment by buldre** [comment]
Does it matter which name you use for the credentials file? I have not seen it starting with an L before.

Did you check the file privileges for the credentials file?

**Comment by ultima-originem** [comment]
Thanks for pointing that out; this was only a typo in my posting. Here are the real filenames and permissions:

```
pi@dev[~] ll .net*
-rwxrwxrwx 1 pi pi 189 Jun 23 02:00 .netatmo.credentials*
-rw-rw-r-- 1 pi pi 175 Jun  4 07:00 .netatmo.credentials.2024-06-22
```


The first one is the current one. It originally had the same permissions as the the second one which was in use for years. I changed the permissions after lnetatmo stopped running at 02:00 (as reported).

Note the timestamp on the original file: 07:00 which was the exact time when the code stopped authenticating.

I also changed the permissions to the ones you're using (-rw-r--r-- 1 pi pi 189 Jun 23 02:00 .netatmo.credentials) and ran the standard test to ensure that my code is not responsible:

```
pi@dev[~] /home/pi/miniconda3/bin/python3 /home/pi/miniconda3/lib/python3.9/site-packages/lnetatmo.py
lnetatmo - ERROR: code=400, reason=, body=b'{"error":"invalid_grant"}'
Traceback (most recent call last):
  File "/home/pi/miniconda3/lib/python3.9/site-packages/lnetatmo.py", line 1097, in <module>
    weatherStation = WeatherStationData(authorization)                                # Test DEVICELIST
  File "/home/pi/miniconda3/lib/python3.9/site-packages/lnetatmo.py", line 416, in __init__
    self.getAuthToken = authData.accessToken
  File "/home/pi/miniconda3/lib/python3.9/site-packages/lnetatmo.py", line 241, in accessToken
    if self.expiration < time.time() : self.renew_token()
  File "/home/pi/miniconda3/lib/python3.9/site-packages/lnetatmo.py", line 252, in renew_token
    if self.refreshToken != resp['refresh_token']:
TypeError: 'NoneType' object is not subscriptable
```

Any more ideas/tests?

**Comment by rvk01** [comment]
> The first one is the current one. It originally had the same permissions as the the second one which was in use for years. I changed the permissions after lnetatmo stopped running at 02:00 (as reported).
> 
> Note the timestamp on the original file: 07:00 which was the exact time when the code stopped authenticating.

If the script could write the file at Jun 23 at 2:00, it should have no problems reading and overwriting it later.
So it can't be a file permission problem.

Either the wrong refresh_token / client_id / client_secret is written to that file... or... netatmo is messing with their systems.

You say you already made a new app (with id and secret etc). Then the data protection officer is also set correctly (I was hoping that that could be the problem if missing).

BTW. You say it stopped working on Jun 23 2:00. Did you mean exactly that time? Because... when the invalid grant happened on 2:00, it shouldn't be writing the credentials to the file (because there aren't any). If the file was written at 2:00 it could crash **3 hours later** (when needing a new refresh-token) OR **at 2:05** when you run a cron-script again and the refresh-token is used again. But not at 2:00.

You say you started fresh at 22:40 with a new refresh token. If you use cron every 5 minutes, then the refresh-token (and every new refresh-token which is received every 5 minutes) did the job correctly for more than 3 hours. So there MUST be something at the end of netatmo where it goes wrong.

Can you build in a sleep of 10 or 30 seconds in your cronjob. I have this:
` */5  *    *   *   *   sleep 10; /home/pi/netatmo_influx.py > /dev/null 2>&1 `

Maybe there is a critical second problem at netatmo where it fails when it's run exactly on a certain time (in combination with that 3 hours where the original refresh token would expire).

(I was also thinking about a possible penalty for not reusing access_token each time consistently, but that would mean everyone should have the problem and here it still runs fine.)

**Comment by ultima-originem** [comment]
You raise some interesting points which I will try to address.

> You say you already made a new app (with id and secret etc). Then the data protection officer is also set correctly (I was hoping that that could be the problem if missing).

I did enter the data protection officer and the app did originally run as expected and when it did, it ran between the following timestamps:
```
FIRST TIMESTAMP: 2024-06-22T20:38:54.000Z
LAST TIMESTAMP:  2024-06-22T23:50:09.000Z
```
which represents a period of 11475s or 3 hours, 11 minutes, and 15 seconds. The timestamp on the credentials file is 9m51s after the last datapoint.

I have changed my crontab to:

`*/5  *     *  *  *  sleep 13; /home/pi/scripts/netatmo.py` 

This fails with the same "invalid_grant" error. I believe I've give you all the relevant data; let me know if I missed something.

**Comment by rvk01** [comment]
> I have changed my crontab to:
> 
> `*/5 * * * * sleep 13; /home/pi/scripts/netatmo.py`
> 
> This fails with the same "invalid_grant" error. I believe I've give you all the relevant data; let me know if I missed something.

Did you create a new refresh_token for that and put it in the credential file?
After that... did it fail directly?

What scopes did you use to create the refresh_token on dev.netatmo.com?

**Comment by ultima-originem** [comment]
I did not create a new refresh toke today; do you want me to do that?
The scope I used was "read_station".

**Comment by philippelt** [comment]
@ultima-originem there is a strange definition of scopes. You may read https://github.com/philippelt/netatmo-api-python/issues/56#issuecomment-1638546613

**Comment by rvk01** [comment]
> I did not create a new refresh token today; do you want me to do that? The scope I used was "read_station".

Yes, you definitely need to create a new one. The old refres_token had expired and won't work anymore.

I used read_station and read_thermostat, but I think read_station should be ok.

**Comment by rvk01** [comment]
> @ultima-originem there is a strange definition of scopes. You may read [#56 (comment)](https://github.com/philippelt/netatmo-api-python/issues/56#issuecomment-1638546613)

Isn't that only if you use homedata (which the default WeatherStationData doesn't use). If you use homedata class, then yes, add those scopes.

**Comment by ultima-originem** [comment]
I have created a new token, using read_station and that results in:
```
pi@dev[~] /home/pi/miniconda3/bin/python3 /home/pi/miniconda3/lib/python3.9/site-packages/lnetatmo.py
lnetatmo - WARNING: Your current token scope do not allow access to Home data
Traceback (most recent call last):
  File "/home/pi/miniconda3/lib/python3.9/site-packages/lnetatmo.py", line 1104, in <module>
    homes = HomeData(authorization)
  File "/home/pi/miniconda3/lib/python3.9/site-packages/lnetatmo.py", line 605, in __init__
    self.rawData = resp['body']
TypeError: 'NoneType' object is not subscriptable

```
This is the first time I'm getting the "Your current token scope do not allow access to Home data" warning. Will now retry with read_station and read_thermostat as @rvk01 did.

Result: no change; I'm still getting the same warning.

Based on the comment by @philippelt I generated a new token with the read_presence scope added.
Result: 

```
pi@dev[~] /home/pi/miniconda3/bin/python3 /home/pi/miniconda3/lib/python3.9/site-packages/lnetatmo.py
lnetatmo - WARNING: No Smokedetectors found
lnetatmo - WARNING: No Cameras found
lnetatmo - WARNING: No Persons found
lnetatmo - WARNING: No events found
lnetatmo - WARNING: No thermostat avaible for testing
lnetatmo - WARNING: Your current token scope do not allow access to HomeCoach
Traceback (most recent call last):
  File "/home/pi/miniconda3/lib/python3.9/site-packages/lnetatmo.py", line 1124, in <module>
    Homecoach = HomeCoach(authorization)
  File "/home/pi/miniconda3/lib/python3.9/site-packages/lnetatmo.py", line 955, in __init__
    self.rawData = resp['body']['devices']
TypeError: 'NoneType' object is not subscriptable
```
The response with the new token and scope is fundamentally different from the responses before today, when I always got "invalid_grant".

Although I get the warnings shown above when I run lnetatmo.py, I noticed approx. 15m later, that my code is now authenticating correctly (with the 3 scopes mentioned earlier). I plan to leave this alone for at least 3hrs. to get past the time where authentication may stop again and report back.

**Comment by ultima-originem** [comment]
@rvk01 This is a misunderstanding which probably originated from the post by @philippelt who asked me to look at [this comment](https://github.com/philippelt/netatmo-api-python/issues/56#issuecomment-1638546613)

Based this comment I expanded the scope from read_station with read_presence and later also with read_thermostat and this is the scope which I'm currently using. This expanded scope generated the warning when running lnetatmo.py but is also delivering data.

What in your opinion should the scope be for reading the weather station and rain data? I can try a reduced scope once I'm past the 3hr. period.

**Comment by rvk01** [comment]
> What in your opinion should the scope be for reading the weather station and rain data? I can try a reduced scope once I'm past the 3hr. period.

I'm using read_station and read_thermostat successfully with just WeatherStationData, with which I read temperature, co2 (inside) and wind and rain (outside modules). But I don't think read_thermostat is needed. Just read_station.

BTW. I see you called lnetatmo.py library directly. I think that might crash on HomeData() if you don't have that. Don't you have your own script which just calls WeatherStationData() ?

**Comment by ultima-originem** [comment]
I tried your scope earlier today and at that time it failed. When I added read_presence it worked but I can't escape the feeling that Netatmo's responses aren't consistent or always predictable. I run my own script and that's working now (for weather station and rain) but also called lnetatmo directly because that's the recommended test in the docs _(from the docs: If you provide all the values in a credential file, you can test that everything is working properly by simply running the package as a standalone program.)_. I don't recall any mentioning of HomeData() in that context.

**Comment by rvk01** [comment]
> I tried your scope earlier today and at that time it failed. When I added read_presence it worked but I can't escape the feeling that Netatmo's responses aren't consistent or always predictable.

Don't use the lnetatmo.py script directly. It has a demo mode where it also calls that homecoach (when called directly) which will fail because of limited scopy. Only use an own script with just a call to WeatherStationData.

**Comment by ultima-originem** [comment]
I agree but would like to note that apparently the docs are misleading/incomplete in this case. Thanks for clearing that up.

**Comment by rvk01** [comment]
Yes. The library could handle the errors returned better (instead of erroring out with that NoneType error). Also note it errored out on HomeCoach so the already called WeatherStationData was successful. But it does show that the library can't handle unexpected results.

**Comment by ultima-originem** [comment]
It's 18:02 here and Netatmo is still authenticating which is very nice but I stil don't know why now and not earlier.

**Comment by philippelt** [comment]
There is obviously an intermittent error on Netatmo side. Others have reach the same conclusion : https://helpcenter.netatmo.com/hc/en-us/community/posts/19320250276626-API-invalid-grant-at-random-intervals?page=2#comments

**Comment by LarsJensen69** [comment]
This error must be at Netatmo. 
I have 2 RPI connected to 2 different Netatmo stations. One in Denmark and one in Spain. 
When this issue occur it is on both of them at the same time +/-2 minutes which is my checking interval. 
They gave absolut nothing in common apart from both authenticating at Netatmo. 
I made a Try-Except-Finally so it reboots automatically every time it fails.

**Comment by LarsJensen69** [comment]
Any chance of getting an update according to the changes described by Netatmo?
https://helpcenter.netatmo.com/hc/en-us/community/posts/20156931057170/comments/20162263407634

**Comment by rvk01** [comment]
> Any chance of getting an update according to the changes described by Netatmo? https://helpcenter.netatmo.com/hc/en-us/community/posts/20156931057170/comments/20162263407634

@LarsJensen69 This has already been implemented in the latest version. The refresh_token is renewed every time and is stored in the credential file.

**Comment by philippelt** [comment]
The initial issue raised can now considered be closed so I am going to lock and close this topic to avoid confusion.
